plotly/presentation_objs/presentation_objs.py

```diff
@@ -0,0 +1,1199 @@
+                )
+            try:
+                var = float(values_dict[key][: -2])
+            except ValueError:
+                raise exceptions.PlotlyError(
+                    NEEDED_STYLE_ERROR_MESSAGE
+                )
+
+        else:
+            var = _percentage_to_pixel(values_dict[key], key)
+        values_dict[key] = var
+
+    return (values_dict['left'], values_dict['top'],
+            values_dict['height'], values_dict['width'])
+
+
+def _remove_extra_whitespace_from_line(line):
+    while line.startswith('\n') or line.startswith(' '):
+        line = line[1:]
+    while line.endswith('\n') or line.endswith(' '):
+        line = line[: -1]
+    return line
+
+
+def _list_of_slides(markdown_string):
+    if not markdown_string.endswith('\n---\n'):
+        markdown_string += '\n---\n'
+
+    text_blocks = re.split(
+        '\n--\n|\n---\n|\n----\n|\n-----\n|\n------\n', markdown_string
+    )
+    list_of_slides = []
+    for j, text in enumerate(text_blocks):
+        if not all(char in ['\n', '-', ' '] for char in text):
+            list_of_slides.append(text)
+
+    if '\n-\n' in markdown_string:
+        msg = ("You have at least one '-' by itself on its own line in your "
+               "markdown string. If you are trying to denote a new slide, "
+               "make sure that the line has 3 '-'s like this: \n\n---\n\n"
+               "A new slide will NOT be created here.")
+        warnings.warn(msg)
+
+    return list_of_slides
+
+
+def _boxes_in_slide(slide):
+    boxes = []
+    slide_copy = copy.deepcopy(slide)
+    prop_split = ';'
+    prop_val_sep = '='
+
+    while '.left' in slide_copy:
+        prop_dict = {}
+        left_idx = slide_copy.find('.left')
+        l_brace_idx = slide_copy[left_idx:].find('{{') + left_idx
+        properties = slide_copy[left_idx + 1: l_brace_idx].split(
+            prop_split
+        )
+
+        # remove white chars from properties
+        empty_props = []
+        for prop in properties:
+            if all(char in [' ', '\n'] for char in prop):
+                empty_props.append(prop)
+
+        for prop in empty_props:
+            properties.remove(prop)
+
+        for prop in properties:
+            prop_name = prop.split(prop_val_sep)[0]
+            prop_val = prop.split(prop_val_sep)[1]
+
+            try:
+                prop_val = float(prop_val)
+            except ValueError:
+                pass
+            prop_dict[prop_name] = prop_val
+
+        r_brace_idx = slide_copy[l_brace_idx:].find('}}') + l_brace_idx
+        box = slide_copy[l_brace_idx + 2: r_brace_idx]
+        box_no_breaks = _remove_extra_whitespace_from_line(box)
+        boxes.append((box_no_breaks, prop_dict))
+
+        slide_copy = slide_copy[r_brace_idx + 2:]
+    return boxes
+
+
+def _top_spec_for_text_at_bottom(text_block, width_per, per_from_bottom=0,
+                                 min_top=30):
+    # TODO: customize this function for different fonts/sizes
```

Comment:
some comments that describe what this function does (inputs and outputs) would be really helpful